File: typescript/src/day_7/crabs.ts
declare type CostFunction = (n: number) => number;

export const fuelCosts = (input: number[]) : number => {
  const mdn = median(input);
  return totalCosts(input, mdn);
}

export const enhancedFuelCosts = (input: number[]) : number => {
  const average = input.reduce((sum, a,i,ar) => {
    sum += a;
    return i==ar.length-1?(ar.length==0?0:sum/ar.length):sum
  },0);
  const mean = Math.floor(average);
  const costFunction = (n: number) => (n * (n + 1) / 2);
  const meanCost = totalCosts(input, mean, costFunction);
  const meanCostP1 = totalCosts(input, mean - 1, costFunction);
  const meanCostM1 = totalCosts(input, mean + 1, costFunction);

  return Math.max.apply(Math, [meanCost, meanCostP1, meanCostM1]);
}

const totalCosts = (
  numbers: number[],
  target: number,
  callback?: CostFunction
) : number => {
  return numbers.reduce((acc, n) => {
    const moves = Math.abs((target - n));

    if(callback && typeof(callback) === "function") {
      return acc += callback(moves);
    }

    return acc += moves;
  }, 0);
}

const median = (numbers: number[]) : number => {
  let middle = Math.floor(numbers.length / 2);
    numbers = [...numbers].sort((a, b) => a - b);
  return numbers.length % 2 !== 0 ? numbers[middle] : (numbers[middle - 1] + numbers[middle]) / 2;
}

File: go/day_07/crabs.go
package main

import (
	"sort"
)

func FuelCount(input []int) int {
	mdn := Median(input)
	return TotalCost(input, mdn)
}

func EnhancedFuelCount(input []int) int {
	mean := Mean(input)
	mean_costs := EnhancedTotalCost(input, mean)
	meanP1_costs := EnhancedTotalCost(input, mean+1)
	meanM1_costs := EnhancedTotalCost(input, mean-1)
	costs := []int{mean_costs, meanP1_costs, meanM1_costs}

	return Min(costs)
}

func TotalCost(numbers []int, target int) int {
	costs := 0
	for i := 0; i < len(numbers); i++ {
		moves := Abs(target - numbers[i])
		costs += moves
	}

	return costs
}

func EnhancedTotalCost(numbers []int, target int) int {
	costs := 0
	for i := 0; i < len(numbers); i++ {
		n := numbers[i]
		moves := Abs(target - n)
		costs += moves + (n * (n + 1) / 2)
	}

	return costs
}

func Median(numbers []int) int {
	sort.Ints(numbers)

	n := len(numbers) / 2

	if n%2 == 0 {
		return (numbers[n-1] + numbers[n]) / 2
	}

	return numbers[n]
}

func Mean(numbers []int) int {
	total := 0

	for _, v := range numbers {
		total += v
	}

	return total / len(numbers)
}

func Abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func Min(numbers []int) int {
	min := numbers[0]
	for _, n := range numbers {
		if n < min {
			min = n
		}
	}
	return min
}
